torchaudio/datasets/tedlium.py

```diff
@@ -0,0 +1,133 @@
+import os
+from typing import Tuple
+
+import torchaudio
+from torch import Tensor
+from torch.utils.data import Dataset
+from torchaudio.datasets.utils import (
+    download_url,
+    extract_archive,
+    walk_files,
+)
+
+RELEASE = "release1"  # Default release
+
+_RELEASE_CONFIGS = {
+    "release1": {
+        "folder_in_archive": "TEDLIUM_release1",
+        "url": "http://www.openslr.org/resources/7/TEDLIUM_release1.tar.gz",
+        "checksum": "ffd31f96d81a21bf4928eaf9bb0b0c2dea7a5247",
+        "data_path": "",
+        "subset": "train",
+    },
+    "release2": {
+        "folder_in_archive": "TEDLIUM_release2",
+        "url": "http://www.openslr.org/resources/19/TEDLIUM_release2.tar.gz",
+        "checksum": "5c8fb045246d1c64296f57b47aa7dc79d16b184f",
+        "data_path": "",
+        "subset": "train",
+    },
+    "release3": {
+        "folder_in_archive": "TEDLIUM_release-3",
+        "url": "http://www.openslr.org/resources/51/TEDLIUM_release-3.tgz",
+        "checksum": "685d27c39c53217383d7933cc405a07048004127",
+        "data_path": "data/",
+        "subset": None,
+    },
+}
+
+
+def load_tedlium_item(
+    fileid: str, line: int, path: str, ext_audio: str, ext_txt: str
+) -> Tuple[Tensor, int, str, int, int, int]:
+    transcript_path = os.path.join(path, "stm/", fileid)
+    with open(transcript_path + ext_txt) as f:
+        transcript = f.readlines()[line]
+        talk_id, _, speaker_id, start_time, end_time, identifier, transcript = transcript.split(
+            " ", 6
+        )
+
+    wave_path = os.path.join(path, "sph/", fileid)
+    waveform, sample_rate = torchaudio.load(wave_path + ext_audio)
```

Comment:
Yes, makes sense

Reply:
@mthrok -- What's the advantage of doing this over having `torchaudio.load` inside the load item function? Is your suggestion to also eventually have a `load_text` method? One property of having a single load item function was that it factored the item loading as a pure function that was taking an identifier to return a data point.

Reply:
> @mthrok -- What's the advantage of doing this over having `torchaudio.load` inside the load item function? Is your suggestion to also eventually have a `load_text` method? One property of having a single load item function was that it factored the item loading as a pure function that was taking an identifier to return a data point.

As far as I understood it he wants to give the user the option to overwrite the load function of the dataset to use optional parameters if he feels like.

So user could do:
```python
def custom_load(path):
    return torchaudio.load(path,normalize=False,channels_first=False)
dt = torchaudio.datasets.TEDLIUM()
dt.load_audio = custom_load
```
 
As `normalize` and `channels_first` are True by default and the `load_item` function doesnt change them the user has no ability to change them.

Reply:
Another example is Windows users who cannot use `torchaudio.load` because torchaudio on Windows does not support `sph` format so they need to provide something else in the custom load method.

Reply:
> As `normalize` and `channels_first` are True by default and the `load_item` function doesnt change them the user has no ability to change them.

If this is indeed something desired, we could also consider exposing those parameters to the constructor instead. I wouldn't say this is enough alone to change the implementation through the addition of a load method.

In torchaudio, we went with the convention that waveforms are between -1 and 1 with batch first, see [readme](https://github.com/pytorch/audio/blob/master/README.md). This is in particular to avoid having to carry such parameters everywhere.

> Another example is Windows users who cannot use `torchaudio.load` because torchaudio on Windows does not support `sph` format so they need to provide something else in the custom load method.

This is an interesting point, and there's precedence for this for datasets using mp3. I see two paths for this:

* Offering a backend that supports `sph` in windows
* Decoupling `torchaudio.load` from `torchaudio.datasets`.

In the former, a user who decides to change the datasets would copy-paste the code and modify to their heart's content. In the latter, we need to decide how we would standardize this. Is any of torch{text,vision} doing something like this already? Thoughts?

Reply:
I would say offering support for sph in windows would be better, less hassling and moving around for the user. At the end, the users who use this custom dataset loaders are the ones looking for a working solution and easy to implement, if they want to customize loading or whatever they will probably rewrite a dataset on their own.